File: desktop/USBDeviceSwift/RaceflightControllerHIDExample/RaceflightControllerHIDExample/maxmix_comm/KBCommService.swift
//
//  KBCommService.swift
//  RaceflightControllerHIDExample
//
//  Created by Siu Man Hsieh on 2022-04-11.
//  Copyright © 2022 Artem Hruzd. All rights reserved.
//

import Foundation
import USBDeviceSwift

public protocol KBCommServiceDelegate: AnyObject {
    func deviceConnected(_ device: HIDDevice)
    func deviceDisconnected(_ device: HIDDevice)
    func log(_ msg: String)
}

/**
 public func connect(device: HIDDevice)
 public func disconnect()
 public func onAudioSessionsChanged()
 
 dependency:
 sessionProvider: AudioSessionProvider
 deviceMonitor: HIDDeviceMonitor
    deviceMonitor.delegfate.deviceGone
    deviceMonitor.delegfate.deviceDataRead
    deviceMonitor.write
 */
public class KBCommService: NSObject, HIDDeviceMonitorDelegate
{
    static let dummySessionData: SessionData = {
        let space: CChar = Int8(" ".cString(using: .utf8)![0])
        return SessionData(id: SESSION_ID_NULL.uint8,
                           name: (space, space, space, space, space, space, space, space, space, space, space, space, space, space, space, space, space, space, space, space),
                           has_prev: 0,
                           has_next: 0,
                           volume: VolumeData(unknown: 0, volume: 50, isMuted: 0))
    }();
    
    let sessionProvider: AudioSessionProvider

    public weak var delegate: KBCommServiceDelegate?
    
    public private(set) var connectedDevice: HIDDevice?
    private var deviceMonitor: HIDDeviceMonitor?
    
    static private let default_session_id: UInt8 = (SESSION_ID_OUT).uint8

    // keyboard firmware protocoal version.
    // if it's not compatible to ours, we don't send it messages
    private var kb_protocol_version: UInt16 = UInt16(UNKNOWN_PROTOCOL_VERSION) // unknown
    
    // current audio session ID
    private var curr_session_id: UInt8 = default_session_id //default is Out session
    
    private var sendSessionInfoTimer: Timer?
    
    init(sessionProvider: AudioSessionProvider,
         delegate: KBCommServiceDelegate? = nil)
    {
        self.sessionProvider = sessionProvider
        self.delegate = delegate
    }
    
    //
    // MARK: -- Public Methods --
    //
    public func connect(device: HIDDevice, deviceMonitor: HIDDeviceMonitor)
    {
        if let connectedDevice = self.connectedDevice {
            
            if connectedDevice === device {
                return //already connected
            }
            
            // disconnect previously connected device before
            // making new connection
            self.disconnect()
        }
        
        self.connectedDevice = device
        self.deviceMonitor = deviceMonitor
        self.deviceMonitor?.addDelegate(self)
        self.sendProtocolVersion()
        self.delegate?.deviceConnected(device)
    }
    
    public func disconnect()
    {
        self.stopRefreshingSessionInfoTimer()
        if let device = self.connectedDevice
        {
            self.connectedDevice = nil
            self.delegate?.deviceDisconnected(device)
        }
        self.deviceMonitor?.removeDelegate(self)
        self.deviceMonitor = nil
    }
    
    public func onAudioSessionsChanged()
    {
        let newAudioSessions = self.sessionProvider.sessions
        
        // check if current session no longer exists
        if nil == newAudioSessions.firstIndex(where:{ $0.id == curr_session_id })
        {
            // reset audio session to default Output
            curr_session_id = Self.default_session_id
        }

        sendSessionInfo()
        sendSessionDatas()
    }
    
    //
    // MARK: -- Internal Methods --
    //
    private func is_protocol_compatible() -> Bool
    {
        return kb_protocol_version == MAXMIX_PROTOCOL_VERSION
    }

    //every 2 sec, send session_info
    //session_info also serve as a heartbeat
    private func startRefreshingSessionInfoTimer()
    {
        if self.sendSessionInfoTimer == nil
        {
            //send sessionInfo as heartbeat to keyboard every 2 sec
            self.sendSessionInfoTimer = Timer.scheduledTimer(timeInterval: 2,
                                                             target: self,
                                                             selector: #selector(sendSessionInfo),
                                                             userInfo: nil,
                                                             repeats: true)
        }
    }
    
    private func stopRefreshingSessionInfoTimer()
    {
        self.sendSessionInfoTimer?.invalidate()
        self.sendSessionInfoTimer = nil
    }
    
    private func sendProtocolVersion()
    {
        self.sendUInt16Message(command: PROTOCOL_VERSION_EXCHANGE, data: UInt16(MAXMIX_PROTOCOL_VERSION))
    }
    
    @objc func sendSessionInfo()
    {
        let session_count = UInt8(min(Int(UInt8.max), self.sessionProvider.sessions.count))
        let session_info = SessionInfo(count: session_count)
        self.sendStructMessage(command: SESSION_INFO, data: session_info)
    }
    
    private func prevSessionIndex(index sessionIndex: Int) -> Int?
    {
        if (sessionIndex > 0)
        {
            let prevIndex = sessionIndex - 1
            if self.sessionProvider.sessions.count > prevIndex
            {
                return prevIndex
            }
        }

        return nil
    }

    private func nextSessionIndex(index sessionIndex: Int) -> Int?
    {
        let nextIndex = sessionIndex + 1
        if self.sessionProvider.sessions.count > nextIndex
        {
            return nextIndex
        }

        return nil
    }

    private func makeSessionData(index sessionIndex: Int) -> SessionData?
    {
        let session: AudioSession?
        if sessionIndex >= self.sessionProvider.sessions.count {
            session = self.sessionProvider.sessions.last
        }
        else if sessionIndex >= 0 {
            session = self.sessionProvider.sessions[sessionIndex]
        }
        else {
            session = self.sessionProvider.sessions.first
        }
        
        guard let session = session else {
            return nil
        }

        var utf8Name = session.name.utf8CString
            // pack bytes to the size of SessionData.name
        while utf8Name.count < sessionDataNameLen {
            utf8Name.append(0)
        }
        let namePointer: SessionDataName = utf8Name.withUnsafeBytes { bytes in
            return bytes.bindMemory(to: SessionDataName.self)[0]
        }
        
        let sessionData = SessionData(id: session.id,
                                      name: namePointer,
                                      has_prev: prevSessionIndex(index:sessionIndex) != nil ? 1: 0,
                                      has_next: nextSessionIndex(index:sessionIndex) != nil ? 1 : 0,
                                      volume: VolumeData(unknown: 0, volume: 50, isMuted: 0))
        
        return sessionData;
    }
    
    
    public func sendSessionDatas()
    {
        let curr_session_index = self.sessionProvider.sessions.firstIndex(where: { session in
            session.id == self.curr_session_id
        }) ?? 0 // use Output session as default
         
        let curr_session_data: SessionData = makeSessionData(index: curr_session_index) ?? Self.dummySessionData
        self.sendStructMessage(command: CURRENT_SESSION, data: curr_session_data)
        
        if let prev_session_index = prevSessionIndex(index:curr_session_index),
           let prev_session_data = makeSessionData(index: prev_session_index)
        {
            self.sendStructMessage(command: PREVIOUS_SESSION, data: prev_session_data)
        }
        
        if let next_session_index = nextSessionIndex(index:curr_session_index),
           let next_session_data = makeSessionData(index: next_session_index)
        {
            self.sendStructMessage(command: NEXT_SESSION, data: next_session_data)
        }
    }
    
    private func sendStructMessage<StructDataType>(
        command: Command,
        data msgData: StructDataType)
    {
        sendMessage(command: command, data: msgData, serializer: StructSerializer<StructDataType>())
    }

    private func sendUInt16Message(
        command: Command,
        data msgData: UInt16)
    {
        sendMessage(command: command, data: msgData, serializer: UInt16Serializer())
    }

    private func sendMessage<S: IStructSerializer, DataType>(
        command: Command,
        data msgData: DataType,
        serializer: S)
    where DataType == S.MsgDataT
    {
        if (!Thread.isMainThread)
        {
            DispatchQueue.main.async {
                self.sendMessage(command: command,
                                 data: msgData,
                                 serializer: serializer)
            }
            return
        }
        
        assert(Thread.isMainThread)

        guard let device = self.connectedDevice else
        {
            return
        }
        
        var data: Data = Data()
        data.append(0xFD)
        data.append(command.to_command_id())
        serializer.serialize(data: msgData, into: &data)
        
        let str = """
                  Sending cmd=\(command.displayName)
                  """;
        self.delegate?.log(str)
        
        self.deviceMonitor?.write(data, to: device, reportId: 0)
    }
    

    //
    // MARK: -- HIDDeviceMonitorDelegate --
    //
    public func deviceFound(_ device: HIDDevice) {
    }
    
    public func deviceGone(_ device: HIDDevice) {
        if (device.id == self.connectedDevice?.id)
        {
            self.disconnect()
        }
    }

    public func deviceDataRead(_ device: HIDDevice, data: Data, reportId: UInt32) {
        if (device.id != self.connectedDevice?.id) {
            return
        }

        do
        {
            let uint8Deserializer = UInt8Serializer()
            
            // check if the first byte == MSG_ID_PREFIX
            var dataIndex = 0
            let msgPrefix = try uint8Deserializer.deserialize(from: data, index: &dataIndex)
            if msgPrefix == MSG_ID_PREFIX
            {
                let command_id: Command = Command(rawValue: UInt32(try uint8Deserializer.deserialize(from: data, index: &dataIndex)))
                let command_str: String = command_id.displayName
                
                if command_id == PROTOCOL_VERSION_EXCHANGE
                {
                    kb_protocol_version = try UInt16Serializer().deserialize(from: data, index: &dataIndex)
                    if kb_protocol_version != MAXMIX_PROTOCOL_VERSION
                    {
                        self.delegate?.log("Incompatible protocol version. Disconnect.")
                        self.disconnect()
                    }
                    else
                    {
                        self.sendSessionInfo()
                        self.sendSessionDatas()
                        self.startRefreshingSessionInfoTimer()
                    }
                }
                else if is_protocol_compatible()
                {
                    var new_curr_session_id: UInt8 = self.curr_session_id
                    
                    var msgRead: Bool = true
                    switch (command_id)
                    {
                        case CURRENT_SESSION_CHANGED:
                            new_curr_session_id = try uint8Deserializer.deserialize(from: data, index: &dataIndex)
                            break
                        case VOLUME_UP:
                            new_curr_session_id = try uint8Deserializer.deserialize(from: data, index: &dataIndex)
                            break
                        case VOLUME_DOWN:
                            new_curr_session_id = try uint8Deserializer.deserialize(from: data, index: &dataIndex)
                            break
                        case TOGGLE_MUTE:
                            new_curr_session_id = try uint8Deserializer.deserialize(from: data, index: &dataIndex)
                            break
                        case CMD_OK, CMD_ERR:
                            break
                        default:
                            msgRead = false
                            break
                    }

                    if msgRead
                    {
                        if command_id != CMD_OK
                        {
                            let str = """
                                      Reading cmd=\(command_str) curr_session_id=\(String(new_curr_session_id))
                                      """;
                            self.delegate?.log(str)
                        }
                        
                        if (new_curr_session_id != self.curr_session_id)
                        {
                            self.curr_session_id = new_curr_session_id
                            
                            // curr_session_id is changed, so send update
                            // SessionData to keyboard
                            self.sendSessionDatas()
                        }
                    }
                    else // !msgRead
                    {
                        let str = "Unknown message cmd=\(command_str)";
                        self.delegate?.log(str)
                    }
                        
                } // else if is_protocol_compatible()
            }
            else
            {
                let data_str = data.convertByteDataToString()
                let str = """
                          Reading unrecognized data=\(data_str)
                          """;
                self.delegate?.log(str)
                
            }
        }
        catch
        {
            self.delegate?.log("\(error)")
        }

    }
}

File: desktop/monenumpad_desktop/monenumpad_desktop/maxmix/KBCommService.cs
using System;
using System.Timers;
using hidapi;
using monenumpad_desktop.Marshal;
using monenumpad_desktop.Maxmix.structs;

namespace monenumpad_desktop.Maxmix
{

    public interface IKBCommService
    { }

    public class KBCommService : IKBCommService
    {
        private static SessionData dummySessionData = new SessionData(
                                structs.Constants.SESSION_ID_NULL,
                                " ",
                                false,
                                false,
                                new VolumeData(false, false, 50));

        private IAudioSessionProvider sessionProvider;

        private hidapi.Device _connectedDevice;
        public hidapi.Device connectedDevice => _connectedDevice;

        static private byte default_session_id = structs.Constants.SESSION_ID_OUT;

        // keyboard firmware protocoal version.
        // if it's not compatible to ours, we don't send it messages
        private UInt16 kb_protocol_version = structs.Constants.UNKNOWN_PROTOCOL_VERSION; // unknown

        // current audio session ID
        private byte curr_session_id = default_session_id; //default is Out session

        private Timer sendSessionInfoTimer;


        public delegate void KBCommServiceNotifyDelegate(KBCommService kBCommService);
        public event KBCommServiceNotifyDelegate DeviceDisconnected;

        private class OutputOnlySessionProvider : IAudioSessionProvider
        {
            private AudioSession[] _sessions = new AudioSession[] {
                new AudioSession(id: KBCommService.dummySessionData.id,
                                 name: KBCommService.dummySessionData.name)
                
            };

            private int vol = 50;
            private bool isMuted = false;

            AudioSession[] IAudioSessionProvider.sessions => _sessions;

            public int getSessionVolume(byte id)
            {
                return vol;
            }

            public void setSessionVolume(byte id, int vol)
            {
                this.vol = vol;
            }

            public bool isSessionMuted(byte id)
            {
                return this.isMuted;
            }

            public bool toggleMuteSession(byte id)
            {
                this.isMuted = !this.isMuted;
                return this.isMuted;
            }
        }

        public KBCommService(IAudioSessionProvider sessionProvider)
        {
            this.sessionProvider = sessionProvider ?? new OutputOnlySessionProvider();
        }

        //
        // MARK: -- Public Methods --
        //
        public void connect(hidapi.hid_device_info dev)
        {
            if (this.connectedDevice != null) {

                if (this.connectedDevice.id.Value == dev.device_id)
                {
                    return; //already connected
                }

                // disconnect previously connected device before
                // making new connection
                this.disconnect();
            }

            this._connectedDevice = new hidapi.Device();
            this._connectedDevice.Open(dev);
            this._connectedDevice.DataReceived += Device_DataReceived;
            this._connectedDevice.Disconnected += Device_Disconnected;
            this.sendProtocolVersion();
        }

        public void disconnect()
        {
            this.stopRefreshingSessionInfoTimer();
            if (this.connectedDevice != null)
            {
                var device = this.connectedDevice;
                this._connectedDevice.Close();
                this._connectedDevice = null;
                this.DeviceDisconnected?.Invoke(this);
            }
        }

        public bool isConnected {
            get {
                return this._connectedDevice != null;
            }
        }

        public void onAudioSessionsChanged()
        {
            var newAudioSessions = this.sessionProvider.sessions;

            // check if current session no longer exists
            bool currSessionAlive = Array.Exists(newAudioSessions, session => session.id == this.curr_session_id);

            if (!currSessionAlive)
            {
                // reset audio session to default Output
                curr_session_id = KBCommService.default_session_id;
            }

            this.sendSessionInfo();
            this.sendSessionDatas();
        }

        //
        // MARK: -- Internal Methods --
        //
        private bool is_protocol_compatible()
        {
            return this.kb_protocol_version == structs.Constants.MAXMIX_PROTOCOL_VERSION;
        }

        //every 2 sec, send session_info
        //session_info also serve as a heartbeat
        private void startRefreshingSessionInfoTimer()
        {
            if (this.sendSessionInfoTimer == null)
            {
                //send sessionInfo as heartbeat to keyboard every 2 sec
                this.sendSessionInfoTimer = new Timer(2000);
                this.sendSessionInfoTimer.Elapsed += OnTimedEvent;
                this.sendSessionInfoTimer.AutoReset = false;
                this.sendSessionInfoTimer.Enabled = true;
            }
        }

        private void OnTimedEvent(Object source, ElapsedEventArgs e)
        {
            // accessing sendSessionInfoTimer needs to be thread safe
            // because OnTimedEvent is called on a different thread.
            Timer timer = null;
            lock (this)
            {
                timer = this.sendSessionInfoTimer;
            }

            if (timer != null)
            {
                this.sendSessionInfo();
                timer.Enabled = true;
            }
        }

        private void stopRefreshingSessionInfoTimer()
        {
            // accessing sendSessionInfoTimer needs to be thread safe
            // because OnTimedEvent is called on a different thread.
            Timer timer = null;
            lock (this)
            {
                timer = this.sendSessionInfoTimer;
                this.sendSessionInfoTimer = null;
            }

            if (timer != null)
            {
                timer.Stop();
                timer.Dispose();
            }
        }

        private void sendProtocolVersion()
        {
            this.sendUInt16Message(Command.PROTOCOL_VERSION_EXCHANGE, structs.Constants.MAXMIX_PROTOCOL_VERSION);
        }

        private void sendSessionInfo()
        {
            var session_count = Math.Min(Byte.MaxValue, (byte)this.sessionProvider.sessions.Length);
            var session_info = new SessionInfo(session_count);
            this.sendStructMessage(Command.SESSION_INFO, session_info);
        }

        private int? prevSessionIndex(int sessionIndex)
        {
            if (sessionIndex > 0)
            {
                var prevIndex = sessionIndex - 1;
                if (this.sessionProvider.sessions.Length > prevIndex)
                {
                    return prevIndex;
                }
            }

            return null;
        }

        private int? nextSessionIndex(int sessionIndex)
        {
            var nextIndex = sessionIndex + 1;
            if (this.sessionProvider.sessions.Length > nextIndex)
            {
                return nextIndex;
            }

            return null;
        }

        private SessionData? makeSessionData(int sessionIndex)
        {
            AudioSession? session = null;
            if (sessionIndex >= this.sessionProvider.sessions.Length)
            {
                session = this.sessionProvider.sessions.Last();
            }
            else if (sessionIndex >= 0)
            {
                session = this.sessionProvider.sessions[sessionIndex];
            }
            else
            {
                session = this.sessionProvider.sessions.First();
            }

            if (session is AudioSession session1)
            {
                int vol = Math.Max(0, this.sessionProvider.getSessionVolume(session1.id));
                vol = Math.Min(100, vol);
                bool muted = this.sessionProvider.isSessionMuted(session1.id);

                var sessionData = new SessionData(
                                          session1.id,
                                          session1.name,
                                          prevSessionIndex(sessionIndex) != null,
                                          nextSessionIndex(sessionIndex) != null,
                                          new VolumeData(false, muted, (byte)vol));
                return sessionData;
            }
            else
            {
                return null;
            }

        }


        public void sendSessionDatas(bool curr_only = false)
        {
            var curr_session_index = Array.FindIndex(this.sessionProvider.sessions, s => s.id == this.curr_session_id);
            if (curr_session_index == -1)
            {
                curr_session_index = 0;
            }

            SessionData curr_session_data = makeSessionData(curr_session_index) ?? KBCommService.dummySessionData;
            this.log("sendSessionData(curr = {0}). has_prev={1}. has_next={2}.", curr_session_data.name,
                curr_session_data.has_prev, curr_session_data.has_next);
            this.sendStructMessage(Command.CURRENT_SESSION, curr_session_data);
            //this.log(curr_session_data.ToBitVector().ToString());

            if (!curr_only)
            {
                var prev_session_index_nullable = prevSessionIndex(curr_session_index);
                if (prev_session_index_nullable is int prev_session_index)
                {
                    var prev_session_data_nullable = makeSessionData(prev_session_index);
                    if (prev_session_data_nullable is SessionData prev_session_data)
                    {
                        this.log("sendSessionData(prev = {0})", prev_session_data.name);
                        this.sendStructMessage(Command.PREVIOUS_SESSION, prev_session_data);
                    }
                }

                var next_session_index_nullable = nextSessionIndex(curr_session_index);
                if (next_session_index_nullable is int next_session_index)
                {
                    var next_session_data_nullable = makeSessionData(next_session_index);
                    if (next_session_data_nullable is SessionData next_session_data)
                    {
                        this.log("sendSessionData(next = {0})", next_session_data.name);
                        this.sendStructMessage(Command.NEXT_SESSION, next_session_data);
                    }

                }
            } // !curr_only
        }

        private void sendStructMessage<StructDataType>(
            Command command,
            StructDataType msgData) where StructDataType : StructMarshallable
        {
            var bitVector = msgData.ToBitVector();
            var data = bitVector.ToBytes();
            sendMessage(command: command, data: data);
        }

        private void sendUInt16Message(
            Command command,
            UInt16 n)
        {
            byte[] data = FieldMarshalFunc.FromBigEndianUInt16(n);
            sendMessage(command, data);
        }

        private void sendMessage(
            Command command,
            byte[] data)
        {
            //if (!Thread.isMainThread)
            //{
            //    DispatchQueue.main.async {
            //        this.sendMessage(command: command,
            //                                data: msgData,
            //                                serializer: serializer)
            //                    };
            //    return;
            //}
            //
            //assert(Thread.isMainThread)

            if (this.connectedDevice == null)
            {
                return;
            }

            if (command != Command.SESSION_INFO)
            {
                this.log("Sending cmd={0}", command);
            }

            // 1 : MSG_ID_PREFIX 0xFD
            // 1 : command id
            // data
            var msgBytes = new byte[1 + 1 + data.Length];
            msgBytes[0] = structs.Constants.MSG_ID_PREFIX;
            msgBytes[1] = (byte)command;
            data.CopyTo(msgBytes, 2);

            this.connectedDevice.Write(msgBytes, msgBytes.Length);
        }

        void log(string format, params object[] args)
        {
            //TODO
            var msg = String.Format(format, args);
            Console.WriteLine(msg);
        }


        public enum FieldIndex
        {
            MsgIDPrefix, //a byte
            MsgCommand, //a byte
            MsgData
        };
        public readonly static PackedStructLayout<FieldIndex> Layout;

        static KBCommService()
        {
            Layout = new PackedStructLayout<FieldIndex>(new FieldDesc[]{
                        new FieldDesc("MsgIDPrefix", 8), // in bits, 
                        new FieldDesc("MsgCommand", 8), //in bits
                        new FieldDesc("MsgData", 0) //in bits, 0 because we don't know how big the msg data body is
                    });
        }

        //
        // MARK: -- hidapi.Device EventHandler --
        //
        public void Device_Disconnected(object sender, EventArgs e)
        {
            this.log("Device is disconnected.");
            this.disconnect();
        }

        public void Device_DataReceived(object sender, hidapi.DataEventArgs e)
        {
            try
            {
                var data = e.Data;
                var bitVector = new BitVector(data);

                // check if the first byte == MSG_ID_PREFIX
                var msgPrefix = Layout.GetFieldValue(bitVector, FieldIndex.MsgIDPrefix, FieldMarshalFunc.ToByte);
                if (msgPrefix != structs.Constants.MSG_ID_PREFIX) {
                    return;
                }

                var command_id = (Command)Layout.GetFieldValue(bitVector, FieldIndex.MsgCommand, FieldMarshalFunc.ToByte);
                if (command_id == Command.CMD_OK)
                {
                    return;
                }

                this.log("Device_DataReceived({0})", command_id.ToString());

                if (command_id == Command.PROTOCOL_VERSION_EXCHANGE)
                {
                    this.kb_protocol_version = Layout.GetFieldValue(bitVector, FieldIndex.MsgData, FieldMarshalFunc.ToBigEndianUInt16);
                    if (this.kb_protocol_version != structs.Constants.MAXMIX_PROTOCOL_VERSION)
                    {
                        this.log("Incompatible protocol version. Disconnect.");
                        this.disconnect();
                    }
                    else
                    {
                        this.sendSessionInfo();
                        this.sendSessionDatas();
                        this.startRefreshingSessionInfoTimer();
                    }
                }
                else if (is_protocol_compatible())
                {
                    byte new_curr_session_id = this.curr_session_id;
                    bool volChanged = false;

                    bool msgRead = true;
                    switch (command_id)
                    {
                        case Command.CURRENT_SESSION_CHANGED:
                            {
                                new_curr_session_id = Layout.GetFieldValue(bitVector, FieldIndex.MsgData, FieldMarshalFunc.ToByte);
                                break;
                            }
                        case Command.VOLUME_UP:
                            {
                                new_curr_session_id = Layout.GetFieldValue(bitVector, FieldIndex.MsgData, FieldMarshalFunc.ToByte);
                                int new_vol = Math.Min(100, this.sessionProvider.getSessionVolume(new_curr_session_id) + 5);
                                this.sessionProvider.setSessionVolume(new_curr_session_id, new_vol);
                                volChanged = true;
                                break;
                            }
                        case Command.VOLUME_DOWN:
                            {
                                new_curr_session_id = Layout.GetFieldValue(bitVector, FieldIndex.MsgData, FieldMarshalFunc.ToByte);
                                int new_vol = Math.Max(0, this.sessionProvider.getSessionVolume(new_curr_session_id) - 5);
                                this.sessionProvider.setSessionVolume(new_curr_session_id, new_vol);
                                volChanged = true;
                                break;
                            }
                        case Command.TOGGLE_MUTE:
                            {
                                new_curr_session_id = Layout.GetFieldValue(bitVector, FieldIndex.MsgData, FieldMarshalFunc.ToByte);
                                this.sessionProvider.toggleMuteSession(new_curr_session_id);
                                volChanged = true;
                                break;
                            }
                        case Command.CMD_OK:
                        case Command.CMD_ERR:
                            break;
                        default:
                            msgRead = false;
                            break;
                    }

                    if (msgRead)
                    {
                        if (command_id != Command.CMD_OK)
                        {
                            this.log("Reading cmd ={0} curr_session_id ={1}",
                                command_id.ToString(), new_curr_session_id);
                        }

                        if (new_curr_session_id != this.curr_session_id)
                        {
                            this.curr_session_id = new_curr_session_id;

                            // curr_session_id is changed, so send update
                            // SessionData to keyboard
                            this.sendSessionDatas();
                        }
                        else if (volChanged)
                        {
                            this.sendSessionDatas( true );
                        }
                    }
                    else // !msgRead
                    {
                        this.log("Unknown message cmd={0}", command_id.ToString());
                    }

                } // else if is_protocol_compatible()                    
                else
                {
                    var data_str = bitVector.ToString();
                    this.log("Reading unrecognized data ={0}", data_str);
                }
            }
            catch (Exception error)
            {
                this.log("Error: {0}\n", error.Message);
                this.log("Error: {0}\n", error.StackTrace);
            } // try
        } // void deviceDataRead

    } // class KBCommService
} //namespace